docs/design-documents/platform/crash-reporting/crash_reporting.md

```diff
@@ -0,0 +1,276 @@
+
+```
+mbed_error_status_t mbed_error( ... )
+{
+    //Handle the error just as we do now and then do the following to save the context into Crash-Report RAM and reset
+    
+    Read the current Crash Report and calculate CRC
+	If CRC matches what's in Crash-Report RAM: 
+		Update the location with new error information
+		Update Reboot Count
+		Calculate new CRC
+		Update with new CRC value
+	Else (if CRC doesn't match) 
+		//This is the case when we dont have a crash report already stored.
+		Update the location with new error information
+		Set Reboot count to 1
+		Calculate new CRC
+		Update with new CRC value
+
+    Do a system reset //using system_reset() function
+}
+
+```
+The Crash-Report RAM region should also be used for tracking other pieces of information such as the CRC value and the
+auto-reboot count. 
+
+### Mechanism to retrieve and reset the error data after reboot
+
+MbedOS error handling system should implement necessary APIs for application to retrieve and reset the error and/or fault context of the previous fatal error after auto-reboot.
+The below APIs should be implemented.
+
+The below API can be called by application to retrieve the error context captured in the Crash-Report RAM. The error context is copied into the location pointed by *error_info*. Note that the caller should allocate the memory for this location.
+The function should return MBED_ERROR_NOT_FOUND if there is no error context currently stored.
+```C
+//Retrieve the reboot error context
+mbed_error_status_t mbed_get_reboot_error_info(mbed_error_ctx *error_info)
+```
+
+The below API can be called by application to retrieve the fault context captured in the Crash-Report RAM. The error context is copied into the location pointed by *fault_context*. Note that the caller should allocate the memory for this location. Note that the fault context is valid only if the previous reboot was caused by an exception. Whether the previous reboot was caused by an exception can be determined from the error code stored in error context information retrieved using mbed_get_reboot_error_info() API above.
+The function should return MBED_ERROR_NOT_FOUND if there is no fault context currently stored.
+```C
+//Call this function to retrieve the last reboot fault context
+mbed_error_status_t mbed_get_reboot_fault_context (mbed_fault_context_t *fault_context);
```

Comment:
That's correct, there could be Invalid argument(fault == NULL), Item not found because there is no context or MBED_SUCCESS. In addition we can capture the module reporting the error using mbed_error_status_t(in this case its MBED_MODULE_PLATFORM).

Reply:
Even when we return error codes, it is good to have return type as int to avoid type casting have the flexible API. Since error codes are used extensively in storage picked up one example here
https://github.com/ARMmbed/mbed-os/blob/master/features/storage/filesystem/littlefs/LittleFileSystem.cpp#L287